routing_common/transit_types.hpp

```diff
@@ -76,6 +78,26 @@ struct TransitHeader
 
 static_assert(sizeof(TransitHeader) == 32, "Wrong header size of transit section.");
 
+class OsmId
+{
+public:
+  OsmId() = default;
+  explicit OsmId(FeatureId featureId) : m_featureId(featureId) {}
+
+  bool operator==(OsmId const & rhs) const { return m_featureId == rhs.m_featureId; }
+  bool IsEqualForTesting(OsmId const & osmId) const { return *this == osmId; }
+  bool IsValid() const { return m_featureId != kInvalidFeatureId; }
+
+  FeatureId GetFeatureId() const { return m_featureId; }
+
+private:
+  DECLARE_TRANSIT_TYPE_FRIENDS
+  DECLARE_VISITOR_AND_DEBUG_PRINT(OsmId, visitor(m_featureId, "feature_id"))
+
+  // |m_featureId| is a feature id corresponding to osm id which presents the class.
+  FeatureId m_featureId = kInvalidFeatureId;
```

Comment:
Wow, so what's the difference between OsmId and FeatureId, if any? Why OsmId keeps FeatureId and is identical to FeatureId?

Reply:
There's a string field at json with name "osm_id". It contains osm id. This osm id should be converted to feature id before serialization to mwm.

When DeserializerFromJson  meats an OsmId while parsing transit data structure it (DeserializerFromJson) reads the string of osm id, converts the string to uint64_t, convert it to feature id and keep it in OsmId class. Then OsmId class is process regularly at other visitors: Serializer, Deserializer and DebugPrint.

Reply:
Okay, finally I get it. From my point of view, this is quite dangerous and strange, and looks like abuse of visitors. If you need to parse feature id from json and convert it to osm_id, please, either keep two different structures (one with m_featureId field and another with m_osmId field), parse the first one and then via mapping convert it to the second one. Or, write your own Visit() function that will work with only specified visitors. Current approach is very misleading.

Reply:
Ok. I'll try with two different structure. If there will be any problems with it Let's write a special method Visit:

```c++
DECLARE_VISITOR_AND_DEBUG_PRINT(Stop, visitor(m_id, "id"), visitor.VisitOsmId(m_osmId, "osm_id"), ...
```

> If you need to parse feature id from json and convert it to osm_id

On the contrary, I need to parse osm id from json and convert it to feature id.

Reply:
I mean, if you use visitors to parse to/from json, your internal structure must correspond to the json structure. Your internal structure holds osm id. Your json structure holds feature id. IMHO, that's wrong, and misleading for the readers of the code. Any changes in json structure will result in changes in the DECLARE_VISITOR, and, therefore, will affect other potential visitors. So, either move DECLARE_VISITOR to the private part of the class and declare a couple of friends, either create auxiliary structure with m_featureId as a field, and then convert this structure explicitly to the structure with m_osmId. Note, that you may use templates to avoid code duplication, for example:
```c++
template <typename Id>
struct MyStructureWithId
{
  DECLARE_VISITOR(visitor(m_id, IdTraits<Id>::GetName()));

  template <typename OtherId>
  void BuildFrom(MyStructureWithId<OtherId> const & other, map<OtherId, Id> const & mapping)
  {
     copy all fields
     m_id  = mapping[other.m_id];
  }

  ... other fields ...
  Id m_id;
};

Reply:
I added a variant of code which is more understandable I guess, but a little less efficient in term of memory.

> either keep two different structures (one with m_featureId field and another with m_osmId field), parse the first one and then via mapping convert it to the second one.

I added only one class `FeatureIdentifiers` which keep osm id and feature id. And I wrote deserialization from json, serialization and deserialization method for it. PTAL at the last commit.

Reply:
> So, either move DECLARE_VISITOR to the private part of the class and declare a couple of friends

BTW It's done now for all transit structures.

```c++
template <typename Id>
struct MyStructureWithId
{
  DECLARE_VISITOR(visitor(m_id, IdTraits<Id>::GetName()));
...
```

Thanks for idea but I did it without template as I wrote before. I guess my current implementation is not misleading. Please have a look at it.